module/VuFind/src/VuFind/Controller/ApiTrait.php

```diff
@@ -0,0 +1,732 @@
+<?php
+/**
+ * Additional functionality for API controllers.
+ *
+ * PHP version 5
+ *
+ * Copyright (C) The National Library 2015-2016.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:controllers Wiki
+ */
+namespace VuFind\Controller;
+
+/**
+ * Additional functionality for API controllers.
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:controllers Wiki
+ */
+trait ApiTrait
+{
+    /**
+     * Callback function in JSONP mode
+     *
+     * @var string
+     */
+    protected $jsonpCallback = null;
+
+    /**
+     * Whether to pretty-print JSON
+     *
+     * @var bool
+     */
+    protected $jsonPrettyPrint = false;
+
+    /**
+     * Type of output to use
+     *
+     * @var string
+     */
+    protected $outputMode = 'json';
+
+    /**
+     * Available record fields.
+     *
+     * Key is the field name that can be requested. Value is an array containing the
+     * method name to call (either in this class or the record driver) and Swagger
+     * specification fields describing the returned data.
+     *
+     * @var array
+     * @see http://swagger.io/specification/
+     */
+    protected $recordFields = [
```

Comment:
This seems like a strong candidate for something that could be moved to a YAML file. Rather than including this data through a trait, it might make more sense to load it from a config file and inject it into the controllers through a factory. You could use a cache (as we do with searchspecs) to avoid the performance disadvantages of YAML parsing. I think the configuration file approach has a significant advantage since a major weakness of Traits (from the perspective of an application that encourages separation of local customizations) is that you can't modify them without changing core code, and the only way to override them is to extend all of the classes that utilize them. It's best not to put "variable" things in Traits for this reason.